lti_consumer/lti_consumer.py

```diff
@@ -0,0 +1,829 @@
+        part of the POST data. These parameters should not be prefixed.
+        Likewise, The creator of an LTI link can add custom key/value parameters
+        to a launch which are to be included with the launch of the LTI link.
+        In this case, we will automatically add `custom_` prefix before this parameters.
+        See http://www.imsglobal.org/LTI/v1p1p1/ltiIMGv1p1p1.html#_Toc316828520
+        """
+
+        # parsing custom parameters to dict
+        custom_parameters = {}
+        for custom_parameter in self.custom_parameters:
+            try:
+                param_name, param_value = [p.strip() for p in custom_parameter.split('=', 1)]
+            except ValueError:
+                _ = self.runtime.service(self, "i18n").ugettext
+                msg = _('Could not parse custom parameter: {custom_parameter}. Should be "x=y" string.').format(
+                    custom_parameter="{0!r}".format(custom_parameter)
+                )
+                raise LtiError(msg)
+
+            # LTI specs: 'custom_' should be prepended before each custom parameter, as pointed in link above.
+            if param_name not in LTI_PARAMETERS:
+                param_name = 'custom_' + param_name
+
+            custom_parameters[unicode(param_name)] = unicode(param_value)
+
+        return custom_parameters
+
+    @property
+    def is_past_due(self):
+        """
+        Is it now past this problem's due date, including grace period?
+        """
+        due_date = self.due  # pylint: disable=no-member
+        if self.graceperiod is not None and due_date:  # pylint: disable=no-member
+            close_date = due_date + self.graceperiod  # pylint: disable=no-member
+        else:
+            close_date = due_date
+        return close_date is not None and datetime.utcnow() > close_date
+
+    def student_view(self, context):
+        """
+        XBlock student view of this component.
+
+        Makes a request to `lti_launch_handler` either
+        in an iframe or in a new window depending on the
+        configuration of the instance of this XBlock
+
+        Arguments:
+            context (dict): XBlock context
+
+        Returns:
+            xblock.fragment.Fragment: XBlock HTML fragment
+        """
+        fragment = Fragment()
+        loader = ResourceLoader(__name__)
+        context.update(self._get_context_for_template())
+        fragment.add_content(loader.render_mako_template('/templates/html/student.html', context))
+        fragment.add_css(loader.load_unicode('static/css/student.css'))
+        fragment.add_javascript(loader.load_unicode('static/js/jquery.iframeModal.min.js'))
+        return fragment
+
+    @XBlock.handler
+    def lti_launch_handler(self, request, suffix=''):  # pylint: disable=unused-argument
+        """
+        XBlock handler for launching the LTI provider.
+
+        Displays a form which is submitted via Javascript
+        to send the LTI launch POST request to the LTI
+        provider.
```

Comment:
What is the rationale of using a fake form here, as opposed to sending the POST request directly from JavaScript?